Mathlib/Combinatorics/Enumerative/DyckWord.lean

```diff
@@ -196,4 +197,139 @@ lemma two_mul_semilength_eq_length : 2 * p.semilength = p.toList.length := by
+
+lemma count_eq_count_of_take_firstReturn_add_one (h : p.toList ≠ []) :
+    (p.toList.take (p.firstReturn + 1)).count U = (p.toList.take (p.firstReturn + 1)).count D := by
+  have := findIdx_get (w := (length_range p.toList.length).symm ▸ p.firstReturn_lt_length h)
+  simpa using this
+
+lemma count_lt_count_of_lt_firstReturn (i : ℕ) (hi : i < p.firstReturn) :
+    (p.toList.take (i + 1)).count D < (p.toList.take (i + 1)).count U := by
+  have ne := not_of_lt_findIdx hi
+  rw [decide_eq_true_eq, ← ne_eq, get_eq_getElem, getElem_range] at ne
+  exact lt_of_le_of_ne (p.count_D_le_count_U (i + 1)) ne.symm
+
+@[simp]
+lemma firstReturn_add : (p + q).firstReturn = if p = 0 then q.firstReturn else p.firstReturn := by
+  split_ifs with h; · simp [h]
+  rw [← toList_eq_nil, ← ne_eq] at h
+  have u : (p + q).toList = p.toList ++ q.toList := rfl
+  have v := p.firstReturn_lt_length h
+  rw [firstReturn, findIdx_eq]
+  · simp_rw [get_eq_getElem, getElem_range, u, decide_eq_true_eq]
+    constructor
+    · rw [take_append_eq_append_take (l₂ := q.toList),
+        show p.firstReturn + 1 - p.toList.length = 0 by omega,
+        take_zero, append_nil, p.count_eq_count_of_take_firstReturn_add_one h]
+    · intro j hj
+      rw [take_append_eq_append_take, show j + 1 - p.toList.length = 0 by omega,
+        take_zero, append_nil]
+      exact (p.count_lt_count_of_lt_firstReturn j hj).ne'
+  · rw [length_range, u, length_append]
+    exact Nat.lt_add_right _ (p.firstReturn_lt_length h)
+
+@[simp]
+lemma firstReturn_nest : p.nest.firstReturn = p.toList.length + 1 := by
+  have u : p.nest.toList = U :: p.toList ++ [D] := rfl
+  rw [firstReturn, findIdx_eq]
+  · simp_rw [get_eq_getElem, getElem_range, u, decide_eq_true_eq]
+    constructor
+    · rw [take_of_length_le (by simp), ← u, p.nest.count_U_eq_count_D]
+    · intro j hj
+      simp_rw [cons_append, take_cons, count_cons, beq_self_eq_true, ite_true,
+        beq_iff_eq, ite_false, take_append_eq_append_take,
+        show j - p.toList.length = 0 by omega, take_zero, append_nil]
+      have := p.count_D_le_count_U j
+      omega
+  · simp_rw [length_range, u, length_append, length_cons]
+    exact Nat.lt_add_one _
+
+/-- The right part of the Dyck word decomposition. -/
+def rightPart (h : p.toList ≠ []) : DyckWord :=
+  p.drop (p.firstReturn + 1) (p.count_eq_count_of_take_firstReturn_add_one h)
+
+/-- The left part of the Dyck word decomposition. -/
+def leftPart (h : p.toList ≠ []) : DyckWord :=
+  (p.take (p.firstReturn + 1) (p.count_eq_count_of_take_firstReturn_add_one h)).denest
+    (by simp [take, h]) (fun i lb ub ↦ by
+      simp only [take, length_take, lt_min_iff] at ub ⊢
+      replace ub := ub.1
+      rw [take_take, min_eq_left ub.le]
+      rw [show i = i - 1 + 1 by omega] at ub ⊢
+      rw [Nat.add_lt_add_iff_right] at ub
+      exact p.count_lt_count_of_lt_firstReturn _ ub)
+
+@[simp]
+theorem leftPart_nest_add_rightPart (h : p.toList ≠ []) :
+    (p.leftPart h).nest + p.rightPart h = p := by
```

Comment:
WHy is `leftPart` denested but not `rightPart`? I find this confusing

Reply:
This is the exact unambiguous inductive construction of Dyck words (see the last part of #9781).

Reply:
I mean I understand this, but naming them `leftPart` and `rightPart` is misleading in that the reader might expect them to be exactly analogous while there is actually a difference between the two. Can you find better, less, symmetrical, names?

Reply:
I should have answered all your queries.